feat(ui): replace navigation with test explorer

userquin opened the pull request:
### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->
This PR changes the navigation to match the test explorer in VSCode extension:
- remove all vue reactivity from ws client files and file ids, only reactivity in state is required: the server will not handle backpressure on the client and so the current version will hang when there are a lot of files and tests (tested with `test/core`)
- use only 1 detail panel in the navigation with a virtual scroll: included `useVirtualScroll` from VueUse, **we need to use another vue library since the tasks will have children (suite/tests...)** (with children expanded the VueUse virtual scroll will not handle the height properly)
- load files on ws open event, sorting the files received in the initial call, this way we don't need to sort them, sort will be done only once
- added `setInterval` to collect the summary state in `composables/test-state.ts` (this PR will not use `composables/summary.ts`): we need to adjust the timeout, we cannot use `setTimeout`  approach, the server will not handle backpressure and the ui can miss a lot of ticks (check `_collect` function in `composables/test-state.ts`)
- logic to load file children using just the `id` and reload the task when the id changes
- allow `@vitest/ws-client` to change the reactivity based on the data (check `VitestClientOptions` and `createClient`), the current version will use `reactive` for all, that's, files, file ids and state
- filter by status
- TODO: use vue virtual scroller

<!-- You can also add additional context here -->
Maybe we need to move the logic to collect global state to the server, this way the client will just receive the data to be displayed in the ui.

<details>
<summary><i>Second run: the server finished and the client still processing incoming data</i></summary>

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/14e28a02-c517-4c6e-a290-53c4ba2e1e1f)
</details>

<details>
<summary><i>Second run: the CPU 100%, once server finish only 1 core (without dev tools)</i></summary>

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/79a07706-5bf0-4cfb-9e25-04e0fabe9999)
</details>

<details>
<summary><i>Vitest Browser tests (preview provider)</i></summary>

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/d2df6fd8-fe1d-483a-97d2-f61da145f97a)
</details>


### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

JessicaSachs commented on packages/ui/client/composables/client/index.ts:
```diff
@@ -2,33 +2,79 @@ import { createClient, getTasks } from '@vitest/ws-client'
+          const entry = files.value.find(f => f.id === file.id)
+          if (entry) {
+            // callbacks.get({ key: file.id })?.()
+            /* const number = */requestAnimationFrame(() => {
```
Maybe I'm missing some cleanup, but you're not properly tearing down the `requestAnimationFrame` anymore, yeah? Shouldn't we be using `useRequestAnimationFrame` or something from VueUse to ensure we're not leaking?

JessicaSachs reviewed (commented):
I might be missing something wrt the `requestAnimationFrame` loop cleanup. Looks like there's a leak.

userquin commented:
Latest changes:
- updated the components and removed `New**` components (ListView should be also removed)
- `test-state.ts` moved to `summary.ts`
- right now we need to manually start the RAF `collect` loop, I'll try to start/stop when required

There are 2 todos in the code @sheremet-va  :
- since we need to start manually in the code the RAF loop, we need to check if the server is still running, for example a page refresh once finished or while running: check if the code here is fine [checkFinished](https://github.com/vitest-dev/vitest/pull/5854/files#diff-ccf62853bfdf7af25e3724a99331864caf9f10999cb4bb5d1c46156d72b808f3R160-R164)
- when running some test (one file/test/suite) or just filtering and running, the dashboard will show all the files, whereas in the server we only have the test we run: we can switch the behavior to match the server output (we also need to check browser, but should be fine) => [clearResults](https://github.com/vitest-dev/vitest/pull/5854/files#diff-ca4e0265ae5d632ea5bfaeb4a603fe0d5e0992ac248197ae441d867e0bcb3b04R75-R77)

I'm going to include `vue-virtual-scroller` with recycler sroller and [variable size mode](https://github.com/Akryum/vue-virtual-scroller/blob/master/packages/vue-virtual-scroller/README.md#variable-size-mode) (iirc @patak-dev  did a great work at elk.zone)

userquin commented:
Included summary in details panel, the ui tests checking for old layout where we have 4 panels, now we have only 1.

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/f561f051-a02f-491f-b249-1cd1702324d3)

sheremet-va commented on packages/browser/src/client/orchestrator.ts:
```diff
@@ -50,9 +50,14 @@ function createIframe(container: HTMLDivElement, file: string) {
+    await rpcDone()
+    await client.rpc.finishBrowserTests()
+  }
+  finally {
```
Why is this needed? Finishing should be done in `onFinished` hook in UI

userquin replied:
not being called

userquin replied:
Maybe being called without files and with errors, I'm going to check what's happening, maybe the logic must be updated for browser mode.

sheremet-va replied:
Calling `finish` resolves this promise:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/pools/browser.ts#L59

Which in turn resolves this:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/core.ts#L636

Which should always call `onFinished` here:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/core.ts#L655

userquin replied:
`done` being called 3 times (I added a breakpoint)... will `onFinished` being called only once, the first time it is resolved? Then the client can receive some `onTaskUpdate` and the state will be reset to `running` again

userquin replied:
`onFinished` being called only once but we're receiving `onTaskUpdate` after `onFinished` called:

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/30541502-39b4-435c-8905-dddd03bd87ad)

userquin replied:
no idea what's happening but now works again without those changes: I'm going to revert changes

sheremet-va replied:
Looks like it is a bug in Vitest runner - it should wait until all `onTaskUpdate` are resolved before finishing `runTests`

userquin commented:
Now tasks also being filtered:

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/b923ea36-149d-4164-bc25-fc6013e7e14f)

![imagen](https://github.com/vitest-dev/vitest/assets/6311119/8277b06a-1c3a-4a61-aee4-bd1ab55b069b)

userquin commented:
I'm going to submit a new PR with latest changes in main, you can check [this video](https://streamable.com/sk17ed).

These are the pending stuff (can be progressive):
- store filter in session to restore it on page refresh
- add `openedTreeItems`  logic, right now removed in `Explorer.vue`
- add sintax hightlight to the tasks when searching
- allow run tests not just the file (will require some changes in the runner and ws-client)
- show all subtasks when`Only Tests`  is not checked and the files/suites match the search and there are no tests matching the search

To run some ui tests:
1) fork the repo, uncheck `Copy the main branch only`, clone to your local and checkout `[userquin/feat-add-test-explorer-ui](https://github.com/vitest-dev/vitest/tree/userquin/feat-add-test-explorer-ui)`  branch
2 ) build vitest, from root run `pnpm install --frozen-lockfile`
3 ) to run `test/core`, from root folder run `cd test/core && pnpm vitest --ui --open`
4) to run `test/browser`, from root folder run `cd test/browser && nr test:browser:preview`

userquin commented on packages/ui/client/composables/explorer/tree.ts:
```diff
@@ -0,0 +1,594 @@
+        }
+        else {
+          // show the parent if matches the filter or at least one child matches the filter
+          if (match || treeNodes.has(child.id) || filesToShow.has(child.id)) {
```
should check `filesToShow` only for files

userquin commented on packages/ui/client/composables/explorer/tree.ts:
```diff
@@ -0,0 +1,594 @@
+  }
+}
+
+window.taskTree = taskTree
```
remove global and taskTree from window

userquin commented on packages/ui/client/composables/search.ts:
```diff
@@ -16,7 +19,7 @@ const filter = reactive(<Filter>{
   skipped: false,
 })
-// don't remove this computed, reactive filter is not working inside computed filtered
+// don't remove these computed, reactive filter is not working inside computed filtered
```
remove this

userquin replied:
upps, this module will be removed

userquin commented on packages/ui/client/composables/explorer/search.ts:
```diff
@@ -0,0 +1,148 @@
+  skipFilter,
+  successFilter,
+} from './state'
+import type { TreeTaskFilter } from '~/composables/explorer/tree'
```
move all types to `types.ts`

userquin commented on packages/ui/client/composables/explorer/search.ts:
```diff
@@ -0,0 +1,148 @@
+  }
+}
+
+export function recalculateUITreeExplorer() {
```
remove this

userquin commented on packages/ui/client/composables/explorer/search.ts:
```diff
@@ -0,0 +1,148 @@
+      failed: taskTree.summary.filesFailed,
+      success: taskTree.summary.filesSuccess,
+      skipped: taskTree.summary.filesSkipped,
+      running: taskTree.summary.filesRunning,
```
review logic, should be test and no files: move the logic to the tree to collect the state and rename `filteredFiles` back to `filteredTests`

userquin commented on packages/ui/client/composables/explorer/tree.ts:
```diff
@@ -0,0 +1,594 @@
+    this.summary.testsSkipped = data.testsSkipped
+    this.summary.totalTests = data.totalTests
+    // update filtered files
+    filteredFiles.value = this.filteredFiles()
```
try to add new tasks here to the tree: if we have a filter applied while running, all missing subtasks will be loaded once `onFinished` called

userquin commented on packages/ui/client/composables/explorer/search.ts:
```diff
@@ -0,0 +1,148 @@
+    clearTimeout(taskId.value)
+    taskId.value = taskTree.buildNavigationEntries(
+      expandAllFlag,
+      search || failed || skipped || success ? (onlyTests ? showOnlyTests : dontShowOnlyTests) : undefined,
```
use only 1 `showOnlyTests`, just update the value

userquin commented on packages/ui/client/components/Navigation.vue:
```diff
@@ -40,38 +39,43 @@ async function onRunAll(files?: File[]) {
-  files.value.forEach(file => {
+  allExpanded.value = true
+  /*files.value.forEach(file => {
     if (!openedTreeItems.value.includes(file.id)) {
```
add `openedTreeItems` logic again

userquin commented on packages/ui/client/composables/explorer/search.ts:
```diff
@@ -49,30 +44,57 @@ export function useSearch(searchBox: Ref<HTMLDivElement | undefined>) {
+      // defaultShowOnlyTests.showOnlyTests = onlyTests
+      updateFilterStorage(search, failed, success, skipped, onlyTests)
+      explorerTree.filterNodes()
+      // taskId.value = explorerTree.buildNavigationEntries(
```
cleanup

userquin commented on packages/ui/client/composables/explorer/state.ts:
```diff
@@ -29,16 +45,30 @@ export const isFilteredByStatus = computed(() => {
+  return treeFilter.value.expandAll !== false
+
+  // return (noItems && expandAll === false) || expandAll === false/* || (!noItems && expandAll === undefined) */
+})
```
cleanup

userquin commented on packages/ui/client/composables/explorer/filter.ts:
```diff
@@ -0,0 +1,216 @@
+  // if (nodesToExpand.size === 0 && treeFilter.value.expandAll !== false) {
+  //   yield * entries
+  //   return
+  // }
```
clenup

userquin commented:
closing, superseded by  #5907
